lib/Dialect/Pipeline/Transforms/ScheduleLinearPipeline.cpp

```diff
@@ -0,0 +1,140 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions linear pipeline scheduling pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Analysis/DependenceAnalysis.h"
+#include "circt/Analysis/SchedulingAnalysis.h"
+#include "circt/Scheduling/Algorithms.h"
+#include "circt/Scheduling/Utilities.h"
+#include "circt/Support/BackedgeBuilder.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "pipeline-schedule-linear"
+
+using namespace mlir;
+using namespace circt;
+using namespace pipeline;
+
+namespace {
+
+class ScheduleLinearPipelinePass
+    : public ScheduleLinearPipelineBase<ScheduleLinearPipelinePass> {
+public:
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+// Returns true if 'op' should be ignored in the scheduling problem.
+static bool ignoreOp(Operation *op) {
+  return isa<hw::ConstantOp, pipeline::ReturnOp>(op);
+}
+
+void ScheduleLinearPipelinePass::runOnOperation() {
+  auto pipeline = getOperation();
+
+  auto stageOpIt = pipeline.getOps<PipelineStageOp>();
+  auto stageRegOpIt = pipeline.getOps<PipelineStageRegisterOp>();
+
+  if ((std::distance(stageOpIt.begin(), stageOpIt.end()) != 0) ||
+      (std::distance(stageRegOpIt.begin(), stageRegOpIt.end()) != 0)) {
+    pipeline.emitError("Pipeline cannot have any stages or stage registers.");
+    return signalPassFailure();
+  }
+
+  // Load operator info from attribute.
+  auto problem = SharedOperatorsProblem::get(pipeline);
+  auto operatorInfo = getOperatorInfo(pipeline);
+  if (!operatorInfo) {
+    pipeline.emitError() << "Expected 'scheduling.operator_info' attribute on "
+                            "pipeline.";
+    return signalPassFailure();
+  }
+  DenseMap<StringAttr, Problem::OperatorType> operatorTypes;
+  for (auto opInfo : *operatorInfo) {
+    auto opName = opInfo.first;
+    auto opType = problem.getOrInsertOperatorType(opName.strref());
+    problem.setLatency(opType, opInfo.second.latency.getValue().getZExtValue());
+    operatorTypes[opName] = opType;
+  }
+
+  // Set operation operator types.
+  Operation *lastOp = nullptr;
+  for (auto &op : pipeline.getOps()) {
+    // Skip if is a known non-functional operator
+    if (ignoreOp(&op))
+      continue;
+
+    // Lookup operator info.
+    auto opName = op.getName().getIdentifier();
+    auto info = operatorTypes.find(opName);
+    if (info == operatorTypes.end()) {
+      op.emitError() << "Operator info for operation '" << opName
+                     << " not found in 'scheduling.operator_info' "
+                        "attribute.";
+      return signalPassFailure();
+    }
+    problem.insertOperation(&op);
+    problem.setLinkedOperatorType(&op, info->second);
+    lastOp = &op;
+  }
+
+  // Solve!
+  assert(succeeded(problem.check()));
+  if (failed(scheduling::scheduleSimplex(problem, lastOp))) {
```

Comment:
Does this need to use this scheduler, or would a simple ASAP scheduler work? I expected scheduleASAP from the commit message.

Reply:
When I ran the asap scheduler on some example mlir asm, I recall the schedule not being ideal for some reason. Like it was using more registers than necessary by computing the results too early. There is obviously a tradeoff space here. It'd be ideal to have a scheduler which tried to minimize the width of the data between stages.

That reminds mem, @mortbopet, we don't want to be registering constant values like ever. I would suggest just ignoring them entirely when you're constructing the stages.

Reply:
> Like it was using more registers than necessary by computing the results too early.

That's just the ASAP objective, working as intended 😉

> It'd be ideal to have a scheduler which tried to minimize the width of the data between stages.

Agreed, I'll look into that!

As a first step towards such an objective, I can port the lifetime minimization objective from the simplex scheduler paper to our implementation. That doesn't give you exactly what you want, but at least schedules free-floating ops closer to their arguments (or their users) on the critical path.